open-apparel-registry/open-apparel-registry
# Setup project development environment

## Overview

* Use git flow
* Derive from [Operations App Template](https://github.com/azavea/operations-app-template)
* Reconcile two existing repositories ([sourcemap/ursa-minor](https://github.com/sourcemap/ursa-minor) and [sourcemap/ursa-major](https://github.com/sourcemap/ursa-major))
* Vagrant
* Docker Compose service definition
  * NodeJS API based off of [latest LTS](https://github.com/nodejs/Release)
  * React development server
  * MongoDB (try to derive version used based on data dumps)
  * ElasticSearch (try to derive version used)
* Clean up `package.json`
  * See what can be pinned
* STRTA
  * `server`
  * `update`
  * `setup`
* Add `README.md`

Resolves #3 

## Notes

~~We have to use the Rsync shared folder provider because of:~~
* https://github.com/firebase/firebase-js-sdk/issues/1234
* https://github.com/grpc/grpc-node/issues/650

I created a Mapbox API key using the `systems (at) azavea.com` email account.  The key is available in LastPass under `Shared-Civic Apps/Open Apparel Registry Mapbox API key (Development)`.

I couldn't derive any information about the MongoDB or ElasticSearch versions by looking at the dumps, but the application seems to work on the latest versions, so I just pinned the frameworks to their last release on NPM.

It's worth nothing that the API server uses [Nodemon](https://nodemon.io), which seems to depend on Python being available, so I opted to prematurely add a `Dockerfile` to the repository. ~~The API server also seems to depend on tools within `build-essential`.~~

I ran into issues starting the ElasticSearch container which required me to follow [these instructions](
https://www.elastic.co/guide/en/elasticsearch/reference/6.2/docker.html#docker-cli-run-prod-mode) to add an Ansible provisioning step that modifies the `vm.max_map_count` kernel setting.

## Testing Instructions


Provision a `.env` file and add the shared Mapbox API key:

```bash
$ cp .env.template .env
$ nano .env
```

Bring up the development environment:

```bash
$ ./scripts/setup
```

Bring up the Docker Compose services:

```bash
$ vagrant ssh
$ ./scripts/server
```

Visit the application at http://localhost:3000.

Try modifying the frontend and see that the changes are applied after reloading your browser:

```diff
diff --git a/src/react/src/components/LandingAlert.js b/src/react/src/components/LandingAlert.js
index 6fe1151..8cf9bc6 100644
--- a/src/react/src/components/LandingAlert.js
+++ b/src/react/src/components/LandingAlert.js
@@ -24,7 +24,7 @@ class LandingAlert extends PureComponent {
             <div className="alert-wrapper">
               <div className="alert">
                 <p className="heading">Welcome to the Open Apparel Registry (beta)!</p>
-                <p>This tool is undergoing beta testing, however you can already explore over 50,000 facilities around the world. Start your search with the filters on the left. We welcome your feedback as we continue to improve and refine the tool, please email comments to <a href="mailto:info@openapparel.org" style={{ color: 'white' }}>info@openapparel.org</a>.</p>
+                <p>What's up? -- This tool is undergoing beta testing, however you can already explore over 50,000 facilities around the world. Start your search with the filters on the left. We welcome your feedback as we continue to improve and refine the tool, please email comments to <a href="mailto:info@openapparel.org" style={{ color: 'white' }}>info@openapparel.org</a>.</p>
                 <button type="button" className="alert-close" onClick={ this.dismissAlert } aria-label="Close">
                   &times;
                 </button>
```
